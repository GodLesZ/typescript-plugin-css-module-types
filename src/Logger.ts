import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import * as tss from "typescript/lib/tsserverlibrary";
import IOptions from "./IOptions";

export interface ILoggerOptions {
    enabled?: boolean;
    filePath?: string;
}

export default class Logger {
    // @ts-ignore
    private readonly options: ILoggerOptions;
    private readonly tsLogger: tss.server.Logger | null = null;

    private get logPath() {
        const basePath = "/logs/ts-log.log";
        return this.options.filePath
            ? path.join(this.options.filePath, basePath)
            : path.join(os.homedir(), basePath);
    }

    public constructor(tsLogger: tss.server.Logger, options: IOptions | null = null) {
        this.tsLogger = tsLogger;
        this.options = {
            ...(options ? options.log : {})
        };
    }

    public log(message: string) {
        if (!this.options.enabled) {
            return;
        }

        fs.appendFileSync(this.logPath, message + "\n");

        if (this.tsLogger) {
            this.tsLogger.info(`[DEBUG] ${message}\n`);
        }
    }
}
