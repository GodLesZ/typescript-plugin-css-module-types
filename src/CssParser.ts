import path from "path";
import { extractICSS } from "icss-utils";
import nodeSass, { SyncImporter, SyncOptions } from "node-sass";
import postcss from "postcss";
import postcssIcssSelectors from "postcss-icss-selectors";
import IOptions from "./IOptions";
import Logger from "./Logger";

export default class CssParser {
    private readonly processor: postcss.Processor;
    private readonly options: IOptions;
    private readonly logger: Logger;

    public constructor(logger: Logger, options: IOptions) {
        this.logger = logger;
        this.options = options;
        this.processor = postcss([
            postcssIcssSelectors({ mode: "local" }),
        ]);
    }

    public getClasses(scss: string) {
        try {
            const renderedCss = nodeSass.renderSync(this.getSassOptions(scss)).css.toString();
            const processedCss = this.processor.process(renderedCss);
            const rootNode = processedCss.root;

            return extractICSS(rootNode).icssExports;
        } catch (e) {
            this.logger.log(`Error during getClasses: ${e.toString()}`);
            this.logger.log(e.stack);

            return {};
        }
    }

    private getSassOptions(data: string): SyncOptions {
        // @ts-ignore
        const nodeModuleImporter: SyncImporter = (url: string) => {
            if (url[0] === "~") {
                url = path.resolve("node_modules", url.substr(1));
            }

            return { file: url };
        };

        return {
            data,
            importer: nodeModuleImporter,
            ...(this.options.sass && this.options.sass.renderOptions ? this.options.sass.renderOptions : {}),
        };
    }
}
