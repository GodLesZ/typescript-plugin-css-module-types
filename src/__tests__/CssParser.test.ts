import * as fs from "fs";
import * as path from "path";
import CssParser from "../CssParser";
import Logger from "../Logger";

// @ts-ignore
const LoggerMock = jest.fn<Logger, Array<null>>(() => ({
    log: jest.fn(),
    setTsLogger: jest.fn(),
}));

const testFileNames = [
    "test-css.module.css",
    "test-scss.module.scss",
    "empty.module.scss",
];

describe("CssParser", () => {
    testFileNames.forEach(fileName => {
        it(`should get all classes from '${fileName}'`, () => {
            const cssParser = new CssParser(new LoggerMock(), {});

            const fileContent = fs.readFileSync(path.join(__dirname, "fixtures", fileName), "utf8");
            const classes = cssParser.getClasses(fileContent);

            expect(classes).toMatchSnapshot();
        });
    });

    it("should return empty object on error", () => {
        const cssParser = new CssParser(new LoggerMock(), {});
        const classes = cssParser.getClasses(`invalid syntax`);
        expect(classes).toMatchObject({});
    });
});
