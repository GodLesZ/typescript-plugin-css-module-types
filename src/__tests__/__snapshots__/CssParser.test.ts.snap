// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CssParser should get all classes from 'empty.module.scss' 1`] = `Object {}`;

exports[`CssParser should get all classes from 'test-css.module.css' 1`] = `
Object {
  "ClassB": "file__ClassB---2bPVi",
  "class-c": "file__class-c---DZ1TD",
  "classA": "file__classA---2xcnJ",
  "class_d": "file__class_d---1mwNi",
}
`;

exports[`CssParser should get all classes from 'test-scss.module.scss' 1`] = `
Object {
  "child-class": "file__child-class---1mwoB",
  "const": "file__const---1uG5q",
  "default": "file__default---3wY_M",
  "local-class": "file__local-class---3KegX",
  "local-class-2": "file__local-class-2---2h6qz",
  "local-class-inside-global": "file__local-class-inside-global---2xH_Y",
  "local-class-inside-local": "file__local-class-inside-local---QdL6b",
  "nested-class-parent": "file__nested-class-parent---_ft7G",
  "nested-class-parent--extended": "file__nested-class-parent--extended---1642l",
  "reserved-words": "file__reserved-words---rsDKy",
  "section-1": "file__section-1---2EiKX",
  "section-2": "file__section-2---2f4aZ",
  "section-3": "file__section-3---R_Ilj",
  "section-4": "file__section-4---3EjYO",
  "section-5": "file__section-5---1DSe8",
  "section-6": "file__section-6---1RoVP",
  "section-7": "file__section-7---l5yMj",
  "section-8": "file__section-8---3FEWv",
  "section-9": "file__section-9---1TFYE",
}
`;
